lib/std/os/windows.zig

```diff
@@ -1258,6 +1258,7 @@ pub fn WSASocketW(
             .WSAEMFILE => return error.ProcessFdQuotaExceeded,
             .WSAENOBUFS => return error.SystemResources,
             .WSAEPROTONOSUPPORT => return error.ProtocolNotSupported,
+            .WSANOTINITIALISED => return error.NotInitialised,
```

Comment:
Please do the retry logic here, and in any other ws2_32 function call which can return `WSANOTINITIALISED`. The documented errors for WsaStartup are:

```
WSASYSNOTREADY | The underlying network subsystem is not ready for network communication.
-- | --
WSAVERNOTSUPPORTED | The version of Windows Sockets support requested is not provided by this particular Windows Sockets implementation.
WSAEINPROGRESS | A blocking Windows Sockets 1.1 operation is in progress.
WSAEPROCLIM | A limit on the number of tasks supported by the Windows Sockets implementation has been reached.
WSAEFAULT | The lpWSAData parameter is not a valid pointer.
```

In all cases this should map to `error.SystemResources` rather than adding a new error to the `SocketError` set.

Reply:
> @marler8997 thanks for your thoughtful remarks here. I think you raise some valid points, however, I am making the judgement call here that Zig's std lib networking abstraction should not require a special call to WsaStartup for Windows applications to work properly.
> 
> Note that the end-game here is likely to switch to the lower level AFD API and bypass ws2_32 altogether, and we should design the abstractions, APIs, and error sets with that in mind. This WSAStartup thing is unnecessary nonsense having to do with critically flawed API design of ws2_32.dll, but it is not a property of the kernel itself. We can bypass this flaw by using the lower level APIs.

Just to clarify, it is still unclear from Windows-related networking code that I've read if it is possible to bypass initializing winsock2 (WSA); even if we make use of AFD for driving socket syscalls and events.

For projects that currently use AFD (i.e. trio, mio, libuv, etc.), they only use a subset of AFD for polling read/write events for sockets, and use winsock2 for everything else. Using undocumented, _possibly_-unstable portions of AFD that other projects do not use can cause forwards compatibility problems in the case Microsoft chooses to make breaking changes.

As an example, all projects right now rely on `WSASocketW` for initializing sockets. How a socket can be initialized solely through AFD is undocumented, and I haven't seen any code online that demonstrates how it can be done without winsock2. It might even be the case that how sockets are created may rely on startup code that is entirely encapsulated in winsock2, and thus sockets can't be created with AFD alone (perhaps WSAStartup loads ws2_32.dll which then registers something to the Transport Driver Interface (TDI) to then allow sockets to be initialized? who knows...).

For the short-term for Zig's standard library, it might be a good idea to bootstrap networking for Windows with winsock2 and IOCP. In the intermediate-term, it would be good to then do research into prior work and have Zig incrementally adopt AFD for networking.

One thing to also note is that with any of the approaches, inevitably, some sort of global state/initialization is necessary (which as @marler8997 pointed out can be maintained w/ `std.once` or static initialization). This is necessary for:

1. initializing the winsock2 subsystem globally only once,
2. keeping a global pool of AFD handles for polling socket read/write events to scale socket syscalls, and
3. loading function pointers to mandatory syscall extensions on application startup (i.e. `ConnectEx`).

For a good discussion on AFD in general, here's an issue that pries through a lot of AFD's details: https://github.com/python-trio/trio/issues/52